### PR TITLE
set based disk

### PR DESCRIPTION
**Pull request by davitbek**
I made PR for this issue
https://github.com/blade-ui-kit/blade-icons/issues/72

### CODE REVIEW

File: config/blade-icons.php [hunk 1]
```diff
@@ -29,6 +29,18 @@
...
+        //     |-----------------------------------------------------------------
+        //     | Icons disk
+        //     |-----------------------------------------------------------------
+        //     |
+        //     | Provide disk where placed SVG icons. Icons are loaded recursively
+        //     | so there's no need to list every sub-directory.
+        //     |
+        //     */
+        //
+        //     'path' => 'resources/svg',
```
Comment (driesvints) [inline config/blade-icons.php]: This doesn't seems correct?

File: src/Exceptions/CannotRegisterIconSet.php [hunk 2]
```diff
@@ -13,6 +13,11 @@ public static function pathNotDefined(string $set): self
         return new static("The options for the \"$set\" set don't have a path defined.");
     }
 
+    public static function pathOrDiskNotDefined(string $set): self
+    {
+        return new static("The options for the \"$set\" set don't have a path or disk defined.");
```
Comment (driesvints) [inline src/Exceptions/CannotRegisterIconSet.php]:
```suggestion
        return new static("The options for the \"$set\" need provide either a path or disk option.");
```

File: src/Factory.php [hunk 3]
```diff
@@ -68,18 +69,43 @@ public function add(string $set, array $options): self
...
-
-                Blade::component(
-                    SvgComponent::class,
-                    implode('.', array_filter($path + [$file->getFilenameWithoutExtension()])),
-                    $set['prefix']
-                );
+            if (isset($set['path'])) {
+                $this->registerComponentsByPath($set);
+            } else {
+                $this->registerComponentsByDisk($set);
```
Comment (driesvints) [inline src/Factory.php:75]: I'd inverse this and make `path` the default one. That also makes the upgrade path for all first party libraries graceful.

File: src/Factory.php [hunk 4]
```diff
@@ -68,18 +69,43 @@ public function add(string $set, array $options): self
...
             }
         }
     }
 
+    protected function registerComponentsByPath($set) : void
+    {
+        foreach ($this->filesystem->allFiles($set['path']) as $file) {
+            $relativePath = Str::after($file->getPath(), $set['path']);
+            $relativePath = str_replace(DIRECTORY_SEPARATOR, '/', $relativePath);
```
Comment (driesvints) [inline src/Factory.php:84]: Why did you change this?

File: src/Factory.php [hunk 5]
```diff
@@ -68,18 +69,43 @@ public function add(string $set, array $options): self
...
+        $allFiles = Storage::disk($set['disk'])->allFiles();
+        foreach ($allFiles as $file) {
+            $file = str_replace('/', '.', $file);
+            $path = pathinfo($file, PATHINFO_FILENAME);
+
+            Blade::component(
+                SvgComponent::class,
+                $path,
+                $set['prefix']
+            );
```
Comment (driesvints) [inline src/Factory.php]: This component call can be on one line.

File: src/Factory.php [hunk 6]
```diff
@@ -119,6 +147,12 @@ private function getSvgFromPath(string $name, string $path): string
         )));
     }
 
+    private function getSvgFromDisk(string $name, string $disk): string
+    {
+        $path = str_replace('.', DIRECTORY_SEPARATOR, $name) . '.svg';
+        return Storage::disk($disk)->get($path);
```
Comment (driesvints) [inline src/Factory.php:150]:
```suggestion

        return Storage::disk($disk)->get($path);
```

**Review by driesvints — CHANGES_REQUESTED** [review]
Neat! Thanks for the PR. I'll give this an even more thorough review/try-out when you've done the first changes here.

File: config/blade-icons.php (hunk 1 above)
Comment (davitbek) [reply config/blade-icons.php]: Yes true it will be like

'disk' => 'svg-disk'

File: src/Factory.php (hunk 3 above)
Comment (davitbek) [reply src/Factory.php:75]: This case always preference is 'path'
If same time in set given path, and disk then all code will work as before. Disk will be ignore that case

File: src/Factory.php (hunk 4 above)
Comment (davitbek) [reply src/Factory.php:84]: I am using windows.
Subfolder dot notation like <x-icon-solid.camera/> is not work in my side.
This code fixed it.

File: src/Factory.php (hunk 5 above)
Comment (davitbek) [reply src/Factory.php]: Okay

File: src/Factory.php (hunk 6 above)
Comment (davitbek) [reply src/Factory.php:150]: Okay

**Comment by davitbek** [comment]
Hi @driesvints . I have made some changes based your suggestions. Can you check please

**Comment by driesvints** [comment]
FYI: I haven't forgotten about this but I'm probably picking this up as soon as I get Blade UI Kit out of the door so that might still be a while.

File: src/Exceptions/CannotRegisterIconSet.php [hunk 7]
```diff
@@ -15,7 +15,7 @@ public static function pathNotDefined(string $set): self
 
     public static function pathOrDiskNotDefined(string $set): self
     {
-        return new static("The options for the \"$set\" set don't have a path or disk defined.");
+        return new static("The options for the \"$set\" need provide either a path or disk option.");
```
Comment (eballeste) [inline src/Exceptions/CannotRegisterIconSet.php:18]:
```suggestion
        return new static("The options for the \"$set\" need to be provided either a path or disk option.");
```

**Comment by nathan-io** [comment]
Hi, was wondering if this may be merged sometime soon? Would love to use this package, but none of our icon assets are on local disk.

**Comment by driesvints** [comment]
Hey all, first of all I'm so sorry it took me this long to get to this. I took the liberty to send in an alternative implementation which should solve this: https://github.com/blade-ui-kit/blade-icons/pull/111

I still want to thank you for this PR 👍
